File: pushPull.py
"""
This is part of https://github.com/hseltman/pushPull
pushPull.py is the Python version of pushPull.R
See pushPull.R and setup.R for a full overview.

Python details:
One time install: From the operating system prompt run
  'conda install -c conda-forge pysftp'.
  If that fails, try 'pip install pysftp'.

Reference: http://pysftp.readthedocs.io/en/release_0.2.9/cookbook.html
           #pysftp-connection-cd
"""

import pysftp
import os.path


def push(files):
    """ Upload files to sftp server.  See pushPull.R for more details. """
    # Check input
    if isinstance(files, str):
        files = [files]
    if not isinstance(files, (list, tuple)):
        raise TypeError("'files' must be a str or list of str's")
    if not all([isinstance(s, str) for s in files]):
        raise TypeError("all 'files' elements must be str's")

    # Check if files are available
    OK = [os.path.isfile(f) for f in files]
    if not all(OK):
        print("Missing files:",
              ", ".join([f for (f, o) in zip(files, OK) if not o]))
        return "Failed"

    # Get configuration info
    cDir = os.path.expanduser("~/pushPullConfig.csv")
    msg = ("Missing or malformed ~/pushPullConfig.csv",
           "Run:",
           "source('https://raw.githubusercontent.com/hseltman/pushPull/" +
           "master/setup.R'')",
           "setup()",
           "Try again after running setup.")

    try:
        with open(cDir) as fh:
            config = fh.read()
    except FileNotFoundError:
        [print(m) for m in msg]
        return "Failed"

    config = [line.split(',') for line in config.splitlines()]
    if not all(len(e) == 2 for e in config):
        [print(m) for m in msg]
        return "Failed"

    config = {k.strip(): v.strip() for (k, v) in config}
    needed = ('sftpSite', 'sftpName', 'sftpPassword', 'userName')
    missing = [config.get(k) is None for k in needed]
    if any(missing):
        print("Missing config keys:",
              ", ".join([n for (m, n) in zip(missing, needed) if m]))

    # Handle the fact that it is possible that a named folder
    # rather that the starting folder is the first writeable folder.
    split = os.path.split(config['sftpSite'])
    if len(split[0]) == 0:
        site = split[1]
        startDir = None
    else:
        site = split[0]
        startDir = split[1]

    # open the ftp site and change to the writeable folder
    try:
        cnopts = pysftp.CnOpts()
        cnopts.hostkeys = None  # Hard to set on Windows
        sftp = pysftp.Connection(host=site,
                                 username=config['sftpName'],
                                 password=config['sftpPassword'],
                                 cnopts=cnopts)
        try:
            if startDir is not None:
                sftp.cwd(startDir)
        except FileNotFoundError:
            print("the starting folder in 'sftpSite' is wrong")
            return "Failure"
    except pysftp.ConnectionException:
        print("Cannot connect to {}".format(site))
        return "Failure"
    except pysftp.AuthenticationException:
        print("Bad username or password in configuration file")
        return "Failure"

    # Change to user's folder, creating it if needed
    if config['userName'] not in ['', '.']:
        if not sftp.isdir(config['userName']):
            try:
                sftp.mkdir(config['userName'])
            except Exception:
                print("Cannot make folder", config['userName'])
                return "Failure"
        try:
            sftp.chdir(config['userName'])
        except IOError:
            print("Cannot chdir server to", config['userName'])
            return("Failure")

    # push the files
    fail_count = 0
    for f in files:
        try:
            sftp.put(f)
        except IOError:
            print("cannot put {} on server".format(f))
            fail_count += 1
        except OSError:
            print("cannot find {} on your computer".format(f))
            fail_count += 1
    sftp.close()

    if fail_count == len(files):
        return "Failed"
    if fail_count == 0:
        return "Success"
    return "Partial success"


def pull(files, who=None):
    """ Download files from sftp server.  See pushPull.R for more details. """
    # Check input
    if isinstance(files, str):
        files = [files]
    if not isinstance(files, (list, tuple)):
        raise TypeError("'files' must be a str or list of str's")
    if not all([isinstance(s, str) for s in files]):
        raise TypeError("all 'files' elements must be str's")

    # Get configuration info
    cDir = os.path.expanduser("~/pushPullConfig.csv")
    msg = ("Missing or malformed ~/pushPullConfig.csv",
           "Run:",
           "source('https://raw.githubusercontent.com/hseltman/pushPull/" +
           "master/setup.R'')",
           "setup()",
           "Try again after running setup.")

    try:
        with open(cDir) as fh:
            config = fh.read()
    except FileNotFoundError:
        [print(m) for m in msg]
        return "Failure"

    config = [line.split(',') for line in config.splitlines()]
    if not all(len(e) == 2 for e in config):
        [print(m) for m in msg]
        return "Failure"

    config = {k.strip(): v.strip() for (k, v) in config}
    needed = ('sftpSite', 'sftpName', 'sftpPassword', 'userName')
    missing = [config.get(k) is None for k in needed]
    if any(missing):
        print("Missing config keys:",
              ", ".join([n for (m, n) in zip(missing, needed) if m]))

    # Handle the fact that it is possible that a named folder
    # rather that the starting folder is the first writeable folder
    # on the sftp server.  E.g., 'sftpSite' might be 'a.b.c.edu/base', where
    # we need to cd to 'base' to use sftp.
    split = os.path.split(config['sftpSite'])
    if len(split[0]) == 0:
        site = split[1]
        startDir = None
    else:
        site = split[0]
        startDir = split[1]

    # open the ftp site and change to the writeable folder
    try:
        cnopts = pysftp.CnOpts()
        cnopts.hostkeys = None  # Hard to set on Windows
        sftp = pysftp.Connection(host=site,
                                 username=config['sftpName'],
                                 password=config['sftpPassword'],
                                 cnopts=cnopts)
        try:
            if startDir is not None:
                sftp.cwd(startDir)
        except FileNotFoundError:
            print("the starting folder in 'sftpSite' is wrong")
            return "Failure"
    except pysftp.ConnectionException:
        print("Cannot connect to {}".format(site))
        return "Failure"
    except pysftp.AuthenticationException:
        print("Bad username or password in configuration file")
        return "Failure"

    # helper function to convert 'path/file.ext' to 'path/file-user.ext'
    def addUser(f, user):
        user = user.strip()
        if user in ('.', ""):
            user = "instructor"
        (root, ext) = os.path.splitext(f)
        return root + "-" + user + ext

    # pull the files
    fail_count = 0
    for f in files:
        if who is None:
            try:
                sftp.get(f, addUser(f, "instructor"))
            except FileNotFoundError:
                print("'{}' was not found on the server".format(f))
                fail_count += 1
        else:
            with sftp.cd():
                try:
                    sftp.cwd(who)
                except FileNotFoundError:
                    print("'folder {}' is not on the server".format(who))
                    fail_count += 1
                try:
                    sftp.get(f, addUser(f, who))
                except FileNotFoundError:
                    print("'{}' was not found on the server".format(f))
                    fail_count += 1
    sftp.close()

    if fail_count == len(files):
        return "Failed"
    if fail_count == 0:
        return "Success"
    return "Partial success"

File: pushPull.R
# Push/pull project, Jan 2018, H. Seltman
# 
# This is a teaching tool to encourage interaction in programming classes.
#
# It is based on the existance of an FTP server where files are shared and exchanged.
# This code can also be implemented as Python functions, SAS macros, etc.

# Educational use examples:
# 1) Students are working in class and a student wants to know why their code
#    failed or you want to show a working (or not working) example from a
#    particular student.  The student pushes the file with push("myFile.R")
#    and you pull it with pull("myFile.R", who="studentId").  Now you can
#    open the file in your current R session.
# 2) Students are attempting to code something, and you want everyone to
#    "catch up" to your version.  Run push("myfile.R").  Each student can
#    then run pull("myFile.R"), and open your version on their computer.

# Requirements:
# "RCurl" package
# An sftp account (ideally restricted to only sftp, with a filespace limit)

# Details:
#   Students use push(files) to upload file(s) from their computer to
#   the sftp site in their folder (as specified by their setup user name).
#   Typically one or more files in the current folder are specified, but
#   relative and absolute locations are allowed.
#
#   Students use pull(files) to download files to their current folder
#   from the root folder of the ftp site.
#
#   Teachers use push(files) to upload file(s) from their computer to the
#   sftp site's root folder (by specifying user name ".").
#
#   Teachers use pull(files, who="studentUserId") to download one or more
#   student files to their current folder.

# One-time setup usage (per computer user):
#   source("https://raw.githubusercontent.com/hseltman/pushPull/master/setup.R")
#
# One time setup (per R session): source("~/pushPull.R")
#
# As needed usage:
# push(files="myFile.R") copies 'files' from the user's computer to folder
#   "userName" in the ftp folder (where 'userName' comes from the config file).
# pull(files="myFile.R", who=NULL) copies 'files' from the main ftp folder
#   to the user's computer (or from folder 'who'),

# References:
# https://stackoverflow.com/questions/15950396/sftp-get-files-with-r
# https://jonkimanalyze.wordpress.com/2014/11/20/r-quick-sftp-file-transfer/

# Return value for all three functions: invisible NULL

# Upload function
push = function(files) {
  if (length(files) == 0 || !is.character(files))
    stop("'files' must be a string vector")

  # get sftp info from config (or tell user how to run setup)
  fname = path.expand(file.path("~", "pushPullConfig.csv"))
  if (!file.exists(fname) ||
      !is.data.frame((dtf=try(read.csv(fname, as.is=TRUE), silent=TRUE))) ||
      !all(c("key", "value") %in% names(dtf)) ||
      !all(c("sftpSite", "sftpName", "sftpPassword", "userName") %in% dtf$key)) {
    cat("Missing or malformed ~/pushPullConfig.csv\n")
    cat("Run:\n")
    cat("source('https://raw.githubusercontent.com/hseltman/pushPull/master/setup.R'')\n")
    cat("setup()")
    stop("try again after running setup")
  }
  sftpSite = dtf$value[dtf$key=="sftpSite"]
  sftpName = dtf$value[dtf$key=="sftpName"]
  sftpPassword = dtf$value[dtf$key=="sftpPassword"]
  userName = dtf$value[dtf$key=="userName"]

  # Check files
  fid = file.info(files)$isdir
  notFound = is.na(fid)
  if (any(notFound)) {
    stop("missing file(s): '", paste(files[notFound], collapse=", "))
  }
  if (any(fid)) {
    stop("folder not file: ", paste(files[fid], collapse=", "))
  }
  
  # Upload files
  if (!require(RCurl, quietly=TRUE, warn.conflicts=FALSE))
    stop("install 'RCurl' and try again")
  front = paste0("sftp://", sftpName, ":", sftpPassword, "@", sftpSite)
  opts = list(ftp.create.missing.dirs=TRUE)
  for (f in files) {
    fRemote = file.path(userName, basename(f))
    rtn = try(ftpUpload(f, file.path(front, fRemote), .opts=opts), silent=TRUE)
    if (is(rtn, "try-error")) {
      cat("Upload of", f, "failed.\n")
      cat("Message:", as.character(attr(rtn, "condition")))
    }
  }

  invisible(NULL)
}


# Download function
pull = function(files, who=NULL) {
  if (length(files) == 0 || !is.character(files))
    stop("'files' must be a string vector")
  
  # get sftp info from config (or tell user how to run setup)
  fname = path.expand(file.path("~", "pushPullConfig.csv"))
  if (!file.exists(fname) ||
      !is.data.frame((dtf=try(read.csv(fname, as.is=TRUE), silent=TRUE))) ||
      !all(c("key", "value") %in% names(dtf)) ||
      !all(c("sftpSite", "sftpName", "sftpPassword", "userName") %in% dtf$key)) {
    cat("Missing or malformed ~/pushPullConfig.csv\n")
    cat("Run:\n")
    cat("source('https://raw.githubusercontent.com/hseltman/pushPull/master/setup.R'')\n")
    cat("setup()")
    stop("try again after running setup")
  }
  sftpSite = dtf$value[dtf$key=="sftpSite"]
  sftpName = dtf$value[dtf$key=="sftpName"]
  sftpPassword = dtf$value[dtf$key=="sftpPassword"]
  userName = dtf$value[dtf$key=="userName"]
  
  # Fixup files (subdirectories must be through "who")
  files = basename(files)
  if (!is.null(who)) {
    files = paste0(who, "/", files)
  }

  # Download files
  if (!require(RCurl, quietly=TRUE, warn.conflicts=FALSE))
    stop("install 'RCurl' and try again")
  front = paste0("sftp://", sftpName, ":", sftpPassword, "@", sftpSite)
  opts = list(ftp.create.missing.dirs=TRUE)
  for (f in files) {
    url = paste0("sftp://", sftpSite, "/", f)
    userpwd = paste0(sftpName, ":", sftpPassword)
    rtn = try(getURL(url, userpwd=userpwd), silent=TRUE)
    if (is(rtn, "try-error")) {
      cat("Download of", f, "failed.\n")
      cat("Message:", as.character(attr(rtn, "condition")))
    } else {
      fLocal = basename(f)
      rtn = try(write(rtn, fLocal), silent=TRUE)
      if (is(rtn, "try-error")) {
        cat("Download of", fLocal, "succeeded, but save to", getwd(), "failed.\n")
        cat("Message:", as.character(attr(rtn, "condition")))
      }
    }
  }
  
  invisible(NULL)
}
